### DISCUSSION TITLE
gre tunneling in child sa

### DISCUSSION THREAD

**Discussion by joseph1267** [strongswan/strongswan#755] [Q&A]
After the child sa exchange and and new terminal was created
topology:
  [ UE side strongswan ]          [free5gc n3iwf and other core]
10.0.0.8 – 192.168.86.38 === 192.168.86.39 – 10.0.0.1
[gre]60.60.0.3 – 192.168.86.38 === [gre]192.168.86.39 – 192.168.82.5 (n3iwf gtp bind address) ------------ 127.0.0.8 (upf)
at this point, I want to ping from ue 60.60.0.3 to 192.168.82.5 but the ping is not going through.
I checked the xfrm state and policy and it had the same spi for both ways. 60.60.0.3 -> 192.168.82.5 and 192.168.82.5 --> 60.60.0.3 in free5gc, but in ue strongswan it had different spi for each ways. I think this is the many reason the ping is not working, but I am not sure.
when the pdu session establishment request message is sent from the ue, free5gc will receive the message and start a Create Child SA
In the proposal of the new SA, it will allocate a new spi.
UE side strongswan will receive the message and start the create child sa
strongswan will use the spi that was sent as the spi for the free5gc and will allocate a new spi for itself and send it through the response SA
strongswan will use the spi sent from free5gc for strongswan -> free5gc xfrm state spi and will use new strongswan’s spi for free5gc -> strongswan xfrm state spi.
when the free5gc receives the IKE message, it will use the strongswan’s spi to generate the key and apply the xfrm state spi.
As a result spi and the key miss match comes out. 
I tried changing the xfrm state and policy in the strongswan side to match with the free5gc and the other way around and both did not work. The interface is using a key 9 for gre interface and the xfrm is created 9 for strongswan but 5 for free5gc (free5gc code reason) and I also suspect this could be a problem for the ping not going from one way or the other. If there is an error in my acknowledgement please guide me to the right way.
Ideal:
[esp][gre][packets] ->>>> to free5gc and other way around

**Comment by tobiasbrunner** [comment]
Sorry, I don't understand your description. What messages are exchanged exactly? What SPIs are exchanged where? Why do they end up to be the same on the other end? And why don't you report this behavior to the free5gc devs? (It seems to be a bug there, or at least some kind of weird limitation.)

**Reply by joseph1267** [reply]
Sorry for the confusion, I did asked the query in the free5gc too...
Create Child SA IKE message is being exchanged. 
Free5gc is sending the Create Child SA request with allocated new SPI to the strongswan
Strongswan also replies with the Create Child SA response with new SPI to the free5gc and both will establish the new Tunnel with the new SPI. And yes I think that is a bug in the free5gc and I have changed the xfrm policy and state to match the one that strongswan did, but still the ping was not going through the tunnel. As far as I understand, when xfrm policy and state matches the other end with the correct interface routing, ping should go through the other side with the correct esp encapsulation... What can be the error here?
Sorry again for the confusion.

**Reply by tobiasbrunner** [reply]
User error? It's tricky to get this right manually.
